tests/testapps/AddressBookSample/AddressBook/SingleContactViewController.m

```diff
@@ -0,0 +1,417 @@
+ */
+- (void)_modifyContact;
+
+/**
+ * Creates a new contact using the values in the text fields, and then adds
+ * said contact to the AddressBook. The phone number is assigned with the label of
+ * Main Phone, and the email is assigned with a label of Work. The address creates
+ * an address with the given value as the street, with a label of Home.
+ */
+- (void)_createContact;
+
+/**
+ * Returns the value in the text field of the given UITextField*, or
+ * NULL if the text field has no text in it.
+ */
+- (CFTypeRef)_copyContactInfoFrom:(UITextField*)textField;
+
+@end
+
+@implementation SingleContactViewController
+
+- (void)dealloc {
+    // Viewing, modifying, and deleting a contact uses the parent's ABAddressBookRef.
+    // Create, however, has its own, and so it must clean up when finished.
+    if (self.contactOperation == kAddressBookCreateContact) {
+        CFRelease(self.addressBook);
+    }
+}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+
+    // Set delegates for the text fields so that the keyboard can be hidden
+    // when return is pressed, or if the user taps outside of the text field.
+    self.firstName.delegate = self;
+    self.lastName.delegate = self;
+    self.phoneNumber.delegate = self;
+    self.email.delegate = self;
+    self.address.delegate = self;
+
+    // Set up the text fields and button depending on operation type.
+    switch (self.contactOperation) {
+        case kAddressBookViewContact:
+            [self _setTextFieldEnabledState:NO];
+            [self.button setTitle:@"Done" forState:UIControlStateNormal];
+            [self _populateTextFields];
+            break;
+        case kAddressBookRemoveContact:
+            [self _setTextFieldEnabledState:NO];
+            [self.button setTitle:@"Remove" forState:UIControlStateNormal];
+            [self _populateTextFields];
+            break;
+        case kAddressBookModifyContact:
+            [self _setTextFieldEnabledState:YES];
+            [self.button setTitle:@"Save" forState:UIControlStateNormal];
+            [self _populateTextFields];
+            break;
+        case kAddressBookCreateContact:
+            [self _setTextFieldEnabledState:YES];
+            [self.button setTitle:@"Create" forState:UIControlStateNormal];
+            break;
+    }
+}
+
+- (void)_setTextFieldEnabledState:(BOOL)enabled {
+    [self.firstName setEnabled:enabled];
+    [self.lastName setEnabled:enabled];
+    [self.phoneNumber setEnabled:enabled];
+    [self.email setEnabled:enabled];
+    [self.address setEnabled:enabled];
+}
+
+- (void)_populateTextFields {
+    [self _populateFirstName];
+    [self _populateLastName];
+    [self _populatePhoneNumber];
+    [self _populateEmail];
+    [self _populateAddress];
+    [self _populateContactPhoto];
+}
+
+- (void)_populateFirstName {
+    NSString* firstName = (__bridge_transfer NSString*)ABRecordCopyValue(self.contact, kABPersonFirstNameProperty);
+    if (firstName) {
```

Comment:
The empty String is an allowed first/last name, so length 0 is okay. If both if/else if conditions are not true, the firstName text box will have its default value for the specific operation -- a blank text field that is editable (creating a contact) or non-editable (viewing/removing/modifying a contact).

---
In reply to: [70120288](https://github.com/Microsoft/WinObjC/pull/627#discussion_r70120288) [](ancestors = 70120288)